### PR TITLE
Updates towards stateless softtypes

### PR DESCRIPTION
**Pull request by Affie**
See #783
Fixes for changes in softtype for [dfg #517](https://github.com/JuliaRobotics/DistributedFactorGraphs.jl/pull/517)

- ContinuousEuclid
- prep FactorMetadata

### CODE REVIEW

**Comment by Affie** [comment]
A concern was raised from slack about serialization of ContinuousEuclid{N}, I tested and it still works well:

```juila
julia> DFG.packVariable(fg, v)
Dict{String,Any} with 11 entries:
  "label"          => "a"
  "dataEntry"      => "{}"
  "nstime"         => "0"
  "softtype"       => "ContinuousEuclid{3}"
  "dataEntryType"  => "{}"
  "ppeDict"        => "{}"
  "solverDataDict" => "{\"default\":{\"vecval\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,…
  "smallData"      => "{}"
  "solvable"       => 1
  "tags"           => "[\"VARIABLE\"]"
  "timestamp"      => "2020-07-29T17:49:43.830+02:00"
```

**Comment by GearsAD** [comment]
That's great :+1:

File: src/FactorGraphTypes.jl [hunk 1]
```diff
@@ -132,18 +132,24 @@ $(TYPEDEF)
...
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3,x4::Symbol) = new(x1, x2, x3, x4, nothing, false)
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3,x4::Symbol,x5::Vector{Symbol};dbg::Bool=false) = new(x1, x2, x3, x4, x5, dbg)
+mutable struct FactorMetadata{T}
+  factoruserdata # deprecate, not in use
+  variableuserdata::Union{Vector, Tuple} # deprecate
+  variablesmalldata::Union{Vector, Tuple} # deprecate
+  solvefor::Union{Symbol, Nothing} # Change to Symbol?
+  variablelist::Union{Nothing, Vector{Symbol}} # Vector{Symbol} #deprecate
+  dbg::Bool #
+    cachedata::Union{Nothing,Vector{T}} # New. change to Vector{T}
```
Comment (dehann) [inline src/FactorGraphTypes.jl:142]: might be simpler to use only `::Vector{T}` and drop the Union?  This does work:
```julia
julia> Vector{Nothing}
Array{Nothing,1}
```

This should give better type stability?

File: src/FactorGraphTypes.jl [hunk 2]
```diff
@@ -132,18 +132,24 @@ $(TYPEDEF)
...
-  variablesmalldata::Union{Vector, Tuple}
-  solvefor::Union{Symbol, Nothing}
-  variablelist::Union{Nothing, Vector{Symbol}}
-  dbg::Bool
-  FactorMetadata() = new() # [], []
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3) = new(x1, x2, x3, nothing, nothing, false)
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3,x4::Symbol) = new(x1, x2, x3, x4, nothing, false)
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3,x4::Symbol,x5::Vector{Symbol};dbg::Bool=false) = new(x1, x2, x3, x4, x5, dbg)
+mutable struct FactorMetadata{T}
+  factoruserdata # deprecate, not in use
```
Comment (dehann) [inline src/FactorGraphTypes.jl:136]: Suggest not deprecating this field.  It is likely used in some of the fancier cases like SAS.  See examples in Caesar.

File: src/FactorGraphTypes.jl (hunk 1 above)
Comment (dehann) [reply src/FactorGraphTypes.jl:142]: https://github.com/JuliaRobotics/IncrementalInference.jl/issues/783#issuecomment-665080114

Comment (Affie) [reply src/FactorGraphTypes.jl]: I want to keep Union Nothing around for now. It allows nothing to be used in the vector for other cases. Say you do not want to cache something.

File: src/FactorGraphTypes.jl (hunk 2 above)
Comment (Affie) [reply src/FactorGraphTypes.jl]: I changed comment accordingly, I don't see its usefulness and can't find it populated anywhere. So its worth a look to remove.

**Pull request merged**